packages/router/src/router.tsx

```diff
@@ -1,41 +1,221 @@
+	widths: Record< string, number >;
+	query?: Record< string, any >;
+	params?: Record< string, any >;
+}
+
+export type BeforeNavigate = ( arg: {
+	path: string;
+	query: Record< string, any >;
+} ) => {
+	path: string;
+	query: Record< string, any >;
+};
+
+interface Config {
+	pathArg: string;
+	beforeNavigate?: BeforeNavigate;
+}
+
+export interface NavigationOptions {
+	transition?: string;
+	state?: Record< string, any >;
+}
+
+const RoutesContext = createContext< Match | null >( null );
+export const ConfigContext = createContext< Config >( { pathArg: 'p' } );
+
+const locationMemo = new WeakMap();
+function getLocationWithQuery() {
+	const location = history.location;
+	let locationWithQuery = locationMemo.get( location );
+	if ( ! locationWithQuery ) {
+		locationWithQuery = {
+			...location,
+			query: Object.fromEntries( new URLSearchParams( location.search ) ),
+		};
+		locationMemo.set( location, locationWithQuery );
+	}
+	return locationWithQuery;
+}
 
 export function useLocation() {
 	return useContext( RoutesContext );
 }
 
 export function useHistory() {
-	return useContext( HistoryContext );
+	const { pathArg, beforeNavigate } = useContext( ConfigContext );
+	return useMemo(
+		() => ( {
+			navigate( rawPath: string, options: NavigationOptions = {} ) {
+				const query = getQueryArgs( rawPath );
+				const path = getPath( 'http://domain.com/' + rawPath ) ?? '';
+				const performPush = () => {
+					const result = beforeNavigate
+						? beforeNavigate( { path, query } )
+						: { path, query };
+					return history.push(
+						{
+							search: buildQueryString( {
+								[ pathArg ]: result.path,
+								...result.query,
+							} ),
+						},
+						options.state
+					);
+				};
+
+				/*
+				 * Skip transition in mobile, otherwise it crashes the browser.
+				 * See: https://github.com/WordPress/gutenberg/pull/63002.
+				 */
+				const isMediumOrBigger =
+					window.matchMedia( '(min-width: 782px)' ).matches;
+				if (
+					! isMediumOrBigger ||
+					// @ts-expect-error
+					! document.startViewTransition ||
+					! options.transition
+				) {
+					return performPush();
+				}
+				document.documentElement.classList.add( options.transition );
+				// @ts-expect-error
+				const transition = document.startViewTransition( () =>
+					performPush()
+				);
+				transition.finished.finally( () => {
+					document.documentElement.classList.remove(
+						options.transition ?? ''
+					);
+				} );
+			},
+		} ),
+		[ pathArg, beforeNavigate ]
+	);
 }
 
-export function RouterProvider( { children }: { children: React.ReactNode } ) {
+export default function useMatch(
+	location: LocationWithQuery,
+	routes: Route[],
+	pathArg: string
+): Match {
+	const { query: rawQuery = {} } = location;
+
+	return useMemo( () => {
+		const matcher = new RouteRecognizer();
+		routes.forEach( ( route ) => {
+			matcher.add( [ { path: route.path, handler: route } ], {
+				as: route.name,
+			} );
+		} );
```

Comment:
It's not good that we're creating the `RouteRecognizer` from scratch on each navigation. It could be more constant. Or extracted from the `useMatch` hook and available imperatively on the `history` object.